lxd/lifecycle/instance.go

```diff
@@ -0,0 +1,67 @@
+package lifecycle
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/project"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+)
+
+// Internal copy of the instance interface
+type instance interface {
+	Project() string
+	Operation() *operations.Operation
+	IsSnapshot() bool
+}
+
+// InstanceAction represents a lifecycle event action for instances
+type InstanceAction string
+
+// All supported lifecycle events for instances
+const (
+	InstanceCreated   = InstanceAction("created")
+	InstanceStarted   = InstanceAction("started")
+	InstanceStopped   = InstanceAction("stopped")
+	InstanceShutdown  = InstanceAction("shutdown")
+	InstanceRestarted = InstanceAction("restarted")
+	InstancePaused    = InstanceAction("paused")
+	InstanceResumed   = InstanceAction("resumed")
+	InstanceRestored  = InstanceAction("restored")
+	InstanceDeleted   = InstanceAction("deleted")
+	InstanceRenamed   = InstanceAction("renamed")
+	InstanceUpdated   = InstanceAction("updated")
+)
+
+// Event creates the lifecycle event for an action on an instance
+func (a InstanceAction) Event(kind, name string, inst instance, ctx map[string]interface{}) api.EventLifecycle {
```

Comment:
Please don't use arg type shortcuts like that, please specify each args type. Thanks